lib/streamlit/elements/number_input.py

```diff
@@ -234,14 +234,18 @@ def _number_input(
                 format = "%d" if int_value else "%0.2f"
 
             # Warn user if they format an int type as a float or vice versa.
-            if format in ["%d", "%u", "%i"] and float_value:
+            # todo: Doesn't happen because back message when formatted as an int
+            #       restricts the data sent to be an int.
+            if (
+                ("%d" in format) or ("%u" in format) or ("%i" in format)
+            ) and float_value:
                 import streamlit as st
 
                 st.warning(
                     "Warning: NumberInput value below has type float,"
                     f" but format {format} displays as integer."
                 )
-            elif format[-1] == "f" and int_value:
+            elif "%f" in format and int_value:
```

Comment:
@sfc-gh-rpalmer in the current implementation 
```
st.number_input("Insert a number", 0.5, 36.6, step=0.1, format="%.2f")
``` 
is a valid format string ("%.2f" means two symbols after the dot), I think checking just `"%f" in format` will miss that case

Reply:
I've fixed this.  Now we can handle whatever can be handled by sprintf-js.